### PR TITLE
Implement option for printing custom formats

### PR DESCRIPTION
**Pull request by tmccombs**
This is mostly a proof of concept before merging there is still some work that needs to be done for:

- [x]  Write new tests (both unit tests and integration tests)
- [x] Add documentation to man page
- [x] Potentially add support for outputting arbitrary bytes on unix like we do for normal printing (I'm not going to worry about this for now)
- [x] Benchmark, and optimize if necessary
    - one potential optimizaiton could be to write directly to the stream rather than building an intermediary OsString, I started looking into that, but it got kind of messy.

Also, this doesn't currently support colorized output when using format string. I think that for an initial implementation that is probably fine.

### CODE REVIEW

**Comment by sharkdp** [comment]
Thank you very much for taking a stab at this!

I didn't have time to properly review it, but I think I am in favor of integrating this new option! That is, unless it leads to a performance penalty for the default case (without `--format`).

Something that I thought about in a different context, but seems fitting here: I'm not a huge fan of the `{}`/`{/}`/`{//}`/`{.}`/… syntax, which we borrowed from GNU parallel, I believe. Would it make sense to think about something a bit more verbatim? I think we could easily support something like `{path}`/`{basename}`/`{parent}`/`{without_extension}`/… in addition. What do you think? I'd much rather type out those words than having to look up the syntax each time.

**Comment by tmccombs** [comment]
I like the more verbose names too.

**Comment by sharkdp** [comment]
Irregardless of the short/long syntax discussion, I'd like to see this being integrated to `fd`!

@tmccombs Are you still interested in finishing this? It would probably be best to focus on the benchmarks first, to make sure that we don't run into performance regressions.

**Comment by tmccombs** [comment]
Sorry it took me so long to get back to this. I am interested in finishing this. Though I wouldn't mind some help with benchmarking it.

**Comment by tmccombs** [comment]
I fixed the merge conflicts.

One question I have is if we would rather have a [--printf option](https://github.com/sharkdp/fd/issues/533) with more format optiions, such as access and modified times, owning user and group, etc. Certainly not as extensive as find, but more than just variants of the path. Although, I  really don't love the idea of having two different format syntaxes.

**Comment by sharkdp** [comment]
> One question I have is if we would rather have a [--printf option](https://github.com/sharkdp/fd/issues/533) with more format optiions, such as access and modified times, owning user and group, etc. Certainly not as extensive as find, but more than just variants of the path. Although, I really don't love the idea of having two different format syntaxes.

I would personally like if we extend the `{…}` syntax. For example, we could have `{basename}` as an alias for `{/}`. This would allow us to use `{access_time}` and similar for an extended printing syntax. If we need more control over *how* a field like a date is printed, we could follow Python/Rust f-strings with something like `{access_time:%Y-%m-%d}`. I don't think that we need to have a find `-printf` compatibility mode here. What do you think?

**Comment by tmccombs** [comment]
That sounds good to me.

**Comment by cohml** [comment]
@tmccombs @sharkdp - Any chance this will be merged soon?

I'm dying for this feature and compulsively checking this PR every few days as a result. Please put me out of my misery lol.

File: src/cli.rs [hunk 1]
```diff
@@ -452,6 +452,20 @@ pub struct Opts {
...
+    ///   '{}': path (of the current search result)
+    ///   '{/}': basename
+    ///   '{//}': parent directory
+    ///   '{.}': path without file extension
+    ///   '{/.}': basename without file extension
+    #[arg(
+        long,
+        value_name = "fmt",
+        help = "Print results according to template",
+        conflicts_with = "list_details"
```
Comment (sharkdp) [inline src/cli.rs:465]: :+1:

**Review by sharkdp — APPROVED** [review]
Thank you very much and sorry for the late review!

**Comment by tmccombs** [comment]
I've run a few benchmarks with hyperfine, and haven't noticed a significant difference in performance.

**Pull request merged**